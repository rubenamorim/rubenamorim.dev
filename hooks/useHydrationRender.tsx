import React, { createContext, useState, useEffect, useContext } from 'react';

interface PropTypes {
    children?: React.ReactNode;
}

export const HydrationRenderContext = createContext<boolean>(true);

export const HydrationRenderProvider: React.FC<PropTypes> = (props) => {
    const [isHydrationRender, setIsHydrationRender] = useState(true);

    useEffect(() => {
        setIsHydrationRender(false);
    }, []);

    return (
        <HydrationRenderContext.Provider value={isHydrationRender} {...props} />
    );
};

export const useHydrationRender = (): boolean => {
    return useContext(HydrationRenderContext);
};

export default useHydrationRender;
