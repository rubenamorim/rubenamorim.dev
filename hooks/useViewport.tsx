import {
    useState,
    useCallback,
    useEffect,
    useContext,
    createContext,
} from 'react';

import { useHydrationRender } from './useHydrationRender';

const BREAKPOINTS = {
    extraSmall: '(max-width: 767px)',
    small: '(min-width: 768px) and (max-width: 1023px)',
    medium: '(min-width: 1024px) and (max-width: 1439px)',
    large: '(min-width: 1440px)',
    gteSmall: '(min-width: 768px)',
    gteMedium: '(min-width: 1024px)',
    lteSmall: '(max-width: 1023px)',
    lteMedium: '(max-width: 1439px)',
};

type ViewportContextTypes = Record<keyof typeof BREAKPOINTS, boolean>;

const matchers: {
    name: keyof typeof BREAKPOINTS;
    query: MediaQueryList | undefined;
}[] = Object.entries(BREAKPOINTS).map(([breakpoint, query]) => ({
    name: breakpoint as keyof typeof BREAKPOINTS,
    query:
        typeof window !== 'undefined' ? window?.matchMedia(query) : undefined,
}));

export const ViewportContext = createContext<Partial<ViewportContextTypes>>({});

export const ViewportProvider: React.FC = (props) => {
    const isHydrationRender = useHydrationRender();

    const [breakpoints, setBreakpoints] = useState({});

    const update = useCallback(() => {
        setBreakpoints(
            matchers.reduce(
                (accum, { name, query }) => ({
                    ...accum,
                    [name]: !!query?.matches,
                }),
                {} as ViewportContextTypes
            )
        );
    }, []);

    useEffect(() => {
        update();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isHydrationRender]);

    useEffect(() => {
        matchers.forEach(({ query }) => {
            query?.addEventListener?.('change', update);
        });

        return (): void => {
            matchers.forEach(({ query }) => {
                query?.removeEventListener?.('change', update);
            });
        };
    }, [update]);

    return <ViewportContext.Provider value={breakpoints} {...props} />;
};

export const useViewport = (): Partial<ViewportContextTypes> => {
    return useContext(ViewportContext);
};

export default useViewport;
