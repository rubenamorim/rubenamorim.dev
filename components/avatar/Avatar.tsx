import React, { useRef, useEffect } from 'react';
import cx from 'classnames';

import { useViewport } from 'hooks/useViewport';

interface PropTypes {
    className?: string;
}

const ROTATION_COEFFICIENT = 25;

const Avatar: React.FC<PropTypes> = ({ className }) => {
    const ref = useRef<HTMLImageElement>(null);

    const { lteSmall } = useViewport();

    useEffect(() => {
        if (lteSmall) {
            if (ref.current) {
                ref.current.style.transform = '';
            }

            return;
        }

        const handleMouseMove = (event: MouseEvent) => {
            requestAnimationFrame(() => {
                if (!ref.current) {
                    return;
                }

                const { innerHeight: height, innerWidth: width } = window;
                const { clientX: xVal, clientY: yVal } = event;

                const yRotation =
                    ROTATION_COEFFICIENT * ((xVal - width / 2) / width);
                const xRotation =
                    -ROTATION_COEFFICIENT * ((yVal - height / 2) / height);

                const transform =
                    'perspective(250px) scale(1.1) rotateX(' +
                    xRotation +
                    'deg) rotateY(' +
                    yRotation +
                    'deg)';

                ref.current.style.transform = transform;
            });
        };

        window.addEventListener('mousemove', handleMouseMove);

        return () => {
            window.removeEventListener('mousemove', handleMouseMove);
        };
    }, [lteSmall]);

    return (
        <picture
            ref={ref}
            className={cx(
                'avatar md:scale-110 md:transition-none md:active:scale-100',
                className
            )}
        >
            <source type="image/avif" srcSet="/media/avatar.avif" />
            <source type="image/webp" srcSet="/media/avatar.webp" />
            <source type="image/png" srcSet="/media/avatar.png" />
            <img
                src="/media/avatar.png"
                alt="Ruben's avatar"
                width="421"
                height="421"
            />
        </picture>
    );
};

export default Avatar;
