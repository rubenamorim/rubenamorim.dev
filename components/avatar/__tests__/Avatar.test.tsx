import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import Avatar from '../Avatar';

import { HydrationRenderProvider } from 'hooks/useHydrationRender';
import { ViewportProvider } from 'hooks/useViewport';

const Wrapper: React.FC = ({ children }) => {
    return (
        <HydrationRenderProvider>
            <ViewportProvider>{children}</ViewportProvider>
        </HydrationRenderProvider>
    );
};

describe('Avatar', () => {
    it('should render correctly', () => {
        const { container } = render(<Avatar />, { wrapper: Wrapper });

        expect(screen.getByAltText("Ruben's avatar")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should apply styles when moving the mouse', async () => {
        const { container } = render(<Avatar />, { wrapper: Wrapper });

        fireEvent.mouseMove(container, { clientX: 100, clientY: 100 });

        await waitFor(() => {
            expect(screen.getByAltText("Ruben's avatar")).toHaveStyle(
                'transform: perspective(250px) scale(1.1) rotateX(9.244791666666668deg) rotateY(-10.05859375deg);'
            );
        });
    });
});
