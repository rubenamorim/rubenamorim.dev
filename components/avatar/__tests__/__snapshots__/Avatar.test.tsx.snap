// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render correctly 1`] = `
<div>
  <picture
    class="avatar md:scale-110 md:transition-none md:active:scale-100"
  >
    <source
      srcset="/media/avatar.webp"
      type="image/webp"
    />
    <source
      srcset="/media/avatar.png"
      type="image/png"
    />
    <img
      alt="Ruben's avatar"
      height="421"
      src="/media/avatar.png"
      width="421"
    />
  </picture>
</div>
`;
